src/Core/Framework/App/Url/AppUrlVerifier.php

```diff
@@ -0,0 +1,279 @@
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Lock\Exception\LockAcquiringException;
+use Symfony\Component\Lock\Exception\LockConflictedException;
+use Symfony\Component\Lock\LockFactory;
+use Symfony\Component\Lock\LockInterface;
+use Symfony\Contracts\Cache\CacheInterface;
+use Symfony\Contracts\Cache\ItemInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * @internal
+ */
+#[Package('framework')]
+class AppUrlVerifier
+{
+    public const VERIFICATION_RESULT_CACHE_KEY = 'app_url_verification_result';
+    private const VERIFICATION_CACHE_KEY_PREFIX = 'app_url_verify-';
+    private const BACK_OFF = [
+        1 => 60 * 5,  // 5 minutes
+        2 => 60 * 15, // 15 minutes
+    ];
+    private const MAX_TRIES = 3;
+    private const VERIFY_PATH = '/api/app-system/shop/verify';
+
+    public function __construct(
+        private readonly string $appEnv,
+        private readonly string $shopwareVersion,
+        private readonly CacheItemPoolInterface&CacheInterface $cache,
+        private readonly HttpClientInterface $httpClient,
+        private readonly LockFactory $lockFactory, /** Should be a lock with ttl support - otherwise locks might not expire and verification will not be performed */
+        private readonly ClockInterface $clock = new NativeClock(),
+    ) {
+    }
+
+    public function getCurrentState(): ?VerificationState
+    {
+        $item = $this->cache->getItem(self::VERIFICATION_RESULT_CACHE_KEY);
+
+        if ($item->isHit()) {
+            return $item->get();
+        }
+
+        return null;
+    }
+
+    public function forceVerify(ShopId $shopId): bool
+    {
+        $this->cache->deleteItem(self::VERIFICATION_RESULT_CACHE_KEY);
+
+        return $this->doVerify($shopId, 'app-url-verification-force');
+    }
+
+    public function verify(ShopId $shopId): bool
+    {
+        if ($this->appEnv !== 'prod') {
+            return true;
+        }
+
+        return $this->doVerify($shopId, 'app-url-verification');
+    }
+
+    /**
+     * Finalize verification, check if the given token and key matches what is stored in the cache.
+     */
+    public function completeVerification(string $runId, string $token): bool
+    {
+        $cacheKey = AppUrlVerifier::VERIFICATION_CACHE_KEY_PREFIX . $runId;
+
+        $item = $this->cache->getItem($cacheKey);
+
+        if (!$item->isHit()) {
+            return false;
+        }
+
+        $storedToken = $item->get();
+
+        if (\strlen($storedToken) !== 32 || \strlen($token) !== 32) {
+            return false;
+        }
+
+        return hash_equals($storedToken, $token);
+    }
+
+    private function doVerify(ShopId $shopId, string $lockKey): bool
+    {
+        $appUrl = $shopId->getFingerprint(AppUrl::IDENTIFIER);
+
+        if ($appUrl === null) {
+            return false;
+        }
+
+        $lock = $this->acquireLock($lockKey);
```

Comment:
why do we use locking when we have a unique run-id? or is it to perform the verification only once when it is expired?